File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] vertex) {
        List<List<Integer>> graph = new ArrayList<>();

        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        // 그래프 생성
        for (int[] v : vertex) {
            graph.get(v[0]).add(v[1]);
            graph.get(v[1]).add(v[0]);
        }

        int[] dist = new int[n + 1];
        Arrays.fill(dist, -1);

        Queue<Integer> q = new LinkedList<>();
        q.offer(1);
        dist[1] = 0;

        // BFS
        while (!q.isEmpty()) {
            int cur = q.poll();

            for (int next : graph.get(cur)) {
                if (dist[next] == -1) {
                    dist[next] = dist[cur] + 1;
                    q.offer(next);
                }
            }
        }

        // 최대 거리 찾기
        int max = 0;
        for (int i = 1; i <= n; i++) {
            max = Math.max(max, dist[i]);
        }

        // 개수 세기
        int count = 0;
        for (int i = 1; i <= n; i++) {
            if (dist[i] == max) count++;
        }

        return count;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, vertex) {
    const graph = Array.from({ length: n + 1 }, () => []);

    for (const [a, b] of vertex) {
        graph[a].push(b);
        graph[b].push(a);
    }

    const dist = Array(n + 1).fill(-1);
    const queue = [1];
    dist[1] = 0;

    while (queue.length) {
        const cur = queue.shift();

        for (const next of graph[cur]) {
            if (dist[next] === -1) {
                dist[next] = dist[cur] + 1;
                queue.push(next);
            }
        }
    }

    const max = Math.max(...dist);
    return dist.filter(v => v === max).length;
}